File: algoritms/binarySearch.js
// Step 1: Definition of the binarySearch function
function binarySearch(array, target) {
  // Step 2: Setting the initial indices left and right
  let left = 0;
  let right = array.length - 1;

  // Step 3: Loop until left exceeds right
  while (left <= right) {
    // Step 4: Calculating the middle index
    let mid = Math.floor((left + right) / 2);

    // Step 5: Checking if we found the target value
    if (array[mid] === target) {
      return mid; // Return the index if you find the value
    }

    // Step 6: Shifting the search range
    if (array[mid] < target) {
      // The target value is on the right side
      left = mid + 1;
    } else {
      // The target value is on the left side
      right = mid - 1;
    }
  }

  // Step 7: Return -1 if the element is not found
  return -1;
}

// Example usage:
const sortedArray = [1, 3, 5, 7, 9, 11, 13, 15];
const targetValue = 7;

const resultIndex = binarySearch(sortedArray, targetValue);

// Displaying the result
if (resultIndex !== -1) {
  console.log(`Element found at index: ${resultIndex}`);
} else {
  console.log(`Element not found in the array.`);
}

File: Projects/Testtrack/algoritms/binary_search.py
import math


def binary_search(arr: list, target: int) -> int:
    """
    Performs binary search on a sorted array to find the target value.

    Args:
        arr (list): A sorted list of elements to search through.
        target (int): The value to search for in the array.

    Returns:
        int: The index of the target element if found, -1 otherwise.

    Time Complexity:
        O(log n) where n is the length of the array.
    """
    left_index = 0
    right_index = len(arr) - 1

    while (left_index <= right_index):
        mid = math.floor((left_index + right_index) / 2)

        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left_index = mid + 1
        else:
            right_index = mid - 1

    return -1


if __name__ == "__main__":
    sorted_list = [1, 3, 5, 7, 9, 11, 13, 15]
    target_value = 7

    print(binary_search(sorted_list, target_value))
